### PR TITLE
Remove Powermock from Spring Libraries

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

In preparation for Java 17 in CI Powermock needs to be removed as a testing dependency. Powermock attempts to access JDK internals which are no longer accessible, resulting in either compilation or runtime failures.

To remove Powermock, many changes in this PR were adding package-private mock-able APIs to places where Powermock was used to get around constructor defaults. For example, instead of returning a stubbed instance on instantiation of a type the class now has a package-private API which constructs the type and the testing code, within the same package and sometimes moved to be so, created a sub-class which mocks the API to return the stubbed instance. Mockito wasn't used in this case as it doesn't allow for stubbing of package-private APIs, this could be changed if the mocked class was moved into implementation with protected or public APIs but that's a much riskier change, which wasn't done in this PR.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by saragluna** [comment]
@moarychan FYI, this should also be implemented on the 4.0 branch.

**Comment by saragluna** [comment]
@mrm9084 fyi

**Review by mrm9084 — CHANGES_REQUESTED** [review]

File: sdk/appconfiguration/spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AppConfigurationPropertySource.java [hunk 1]
```diff
@@ -82,7 +82,7 @@
     @Override
     public String[] getPropertyNames() {
         Set<String> keySet = properties.keySet();
-        return keySet.toArray(new String[keySet.size()]);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/appconfiguration/spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AppConfigurationPropertySource.java:85]: Hmm, so both are right. Read through the `toArray(T[])` documentation further and it explains that if the passed array is large enough to fit the elements of the `Collection`, or `Set` in this case, that array will be populated and returned. If the passed array isn't large enough a new one will be allocated internally that is large enough to hold all elements of the `Collection`.

Both will work, but I'll revert this.

**Review by chenrujun — APPROVED** [review]
LGTM.

**Review by mrm9084 — APPROVED** [review]

**Pull request merged**